### ISSUE TITLE
Better `typeKey` support in TypeScript, or drop support for it in TypeScript altogether

### ISSUE THREAD

**Issue by vkarpov15** [Automattic/mongoose#11482]
Re: #9927, #11241

The fact that `type` is optional in `SchemaTypeOptions` causes unpredictable behavior, because incorrect type definitions often end up as valid `SchemaTypeOptions` because of https://github.com/Automattic/mongoose/blob/3dd0461577523f70c149cc662c5dc2f18f05b485/types/index.d.ts#L1582. Something to think about.

**Comment by Uzlopak** [comment]
Yeah, that should be removed imho, as it indicates that the schema type can handle more values. 

Was this added, because than people can add more attributes for their custom types?

Probably instead of having that big T extends bla stuff, we should use type as discriminator.

**Comment by vkarpov15** [comment]
We added it because Mongoose supports schema definitions like this:

```javascript
new Schema({
  name: {
    $type: String // Notice '$type` here instead of `type`... because of `typeKey`
  }
}, { typeKey: '$type' });
```

The original rationale for this was to make it easier to define schemas that have properties named 'type'. Unfortunately, this sort of pattern doesn't work well with TypeScript, and we made `type` optional in SchemaTypeOptions to avoid completely blocking using `typeKey` in TypeScript.

Also, I linked to the wrong line. `[other: string]: any` is correct, because plugins often use additional schema type options. Like the `autopopulate` option from mongoose-autopopulate.

**Comment by vkarpov15** [comment]
We have this in #11487 now, just need some docs.

**Comment by agus-condro-speedoc** [comment]
> We added it because Mongoose supports schema definitions like this:
> 
> ```js
> new Schema({
>   name: {
>     $type: String // Notice '$type` here instead of `type`... because of `typeKey`
>   }
> }, { typeKey: '$type' });
> ```
> 
> The original rationale for this was to make it easier to define schemas that have properties named 'type'. Unfortunately, this sort of pattern doesn't work well with TypeScript, and we made `type` optional in SchemaTypeOptions to avoid completely blocking using `typeKey` in TypeScript.
> 
> Also, I linked to the wrong line. `[other: string]: any` is correct, because plugins often use additional schema type options. Like the `autopopulate` option from mongoose-autopopulate.

Has this been resolved?

In NestJs Typescript declaring the schema's field like that doesn't work, and the recommended way [here](https://mongoosejs.com/docs/guide.html#typeKey) also doesn't work in my VSCode auto suggestions

**Comment by vkarpov15** [comment]
@agus-condro-speedoc docs haven't been done, but everything else has been resolved. Can you provide an example of how you're defining schema fields?

**Comment by vkarpov15** [comment]
We can close this since the `typeKey` option works as expected in TypeScript right now.
